base/linalg/transpose.jl

```diff
@@ -0,0 +1,334 @@
+
+function transpose!(B::AbstractMatrix,A::AbstractMatrix)
+    m, n = size(A)
+    size(B,1) == n && size(B,2) == m || throw(DimensionMismatch("transpose"))
+
+    if m*n<=4*transposebaselength
+        @inbounds begin
+            for j = 1:n #Fixme iter
+                for i = 1:m #Fixme iter
+                    B[j,i] = transpose(A[i,j])
+                end
+            end
+        end
+    else
+        transposeblock!(B,A,m,n,0,0)
+    end
+    return B
+end
+function transpose!(B::AbstractVector, A::AbstractMatrix)
+    length(B) == length(A) && size(A,1) == 1 || throw(DimensionMismatch("transpose"))
+    copy!(B, A)
+end
+function transpose!(B::AbstractMatrix, A::AbstractVector)
+    length(B) == length(A) && size(B,1) == 1 || throw(DimensionMismatch("transpose"))
+    copy!(B, A)
+end
+function transposeblock!(B::AbstractMatrix,A::AbstractMatrix,m::Int,n::Int,offseti::Int,offsetj::Int)
+    if m*n<=transposebaselength
+        @inbounds begin
+            for j = offsetj+(1:n) #Fixme iter
+                for i = offseti+(1:m) #Fixme iter
+                    B[j,i] = transpose(A[i,j])
+                end
+            end
+        end
+    elseif m>n
+        newm=m>>1
+        transposeblock!(B,A,newm,n,offseti,offsetj)
+        transposeblock!(B,A,m-newm,n,offseti+newm,offsetj)
+    else
+        newn=n>>1
+        transposeblock!(B,A,m,newn,offseti,offsetj)
+        transposeblock!(B,A,m,n-newn,offseti,offsetj+newn)
+    end
+    return B
+end
+function ctranspose!(B::AbstractMatrix,A::AbstractMatrix)
+    m, n = size(A)
+    size(B,1) == n && size(B,2) == m || throw(DimensionMismatch("transpose"))
+
+    if m*n<=4*transposebaselength
+        @inbounds begin
+            for j = 1:n #Fixme iter
+                for i = 1:m #Fixme iter
+                    B[j,i] = ctranspose(A[i,j])
+                end
+            end
+        end
+    else
+        ctransposeblock!(B,A,m,n,0,0)
+    end
+    return B
+end
+function ctranspose!(B::AbstractVector, A::AbstractMatrix)
+    length(B) == length(A) && size(A,1) == 1 || throw(DimensionMismatch("transpose"))
+    ccopy!(B, A)
+end
+function ctranspose!(B::AbstractMatrix, A::AbstractVector)
+    length(B) == length(A) && size(B,1) == 1 || throw(DimensionMismatch("transpose"))
+    ccopy!(B, A)
+end
+function ctransposeblock!(B::AbstractMatrix,A::AbstractMatrix,m::Int,n::Int,offseti::Int,offsetj::Int)
+    if m*n<=transposebaselength
+        @inbounds begin
+            for j = offsetj+(1:n) #Fixme iter
+                for i = offseti+(1:m) #Fixme iter
+                    B[j,i] = ctranspose(A[i,j])
+                end
+            end
+        end
+    elseif m>n
+        newm=m>>1
+        ctransposeblock!(B,A,newm,n,offseti,offsetj)
+        ctransposeblock!(B,A,m-newm,n,offseti+newm,offsetj)
+    else
+        newn=n>>1
+        ctransposeblock!(B,A,m,newn,offseti,offsetj)
+        ctransposeblock!(B,A,m,n-newn,offseti,offsetj+newn)
+    end
+    return B
+end
```

Comment:
Boy, there's a lot of near-duplicated code here. I'd really prefer a generic implementation in which you passed `transpose` or `ctranspose` as an argument.